hack/lib/version.sh

```diff
@@ -114,6 +115,16 @@ kube::version::get_version_vars() {
       fi
     fi
   fi
+
+  kube::version::get_kustomize_version_vars
+}
+
+kube::version::get_kustomize_version_vars() {
+  kustomizeTag=$(awk '/sigs.k8s.io\/kustomize\/kustomize\/v4/ {print $4}' go.mod)
```

Comment:
this is clever! I like this approach

Reply:
is the module version available using debug/buildinfo in go1.18 now? I was actually planning to switch as much of the version stamping code to that if possible

Reply:
xref https://pkg.go.dev/runtime/debug

Reply:
That does look preferable! I tried using it by introducing this function:

```go
func getKustomizeVersion() (string, error) {
	info, ok := debug.ReadBuildInfo()
	if !ok {
		return "unknown", fmt.Errorf("unable to read build info to determine Kustomize version")
	}
	for _, dep := range info.Deps {
		if dep.Path == "sigs.k8s.io/kustomize/kustomize/v4" {
			return dep.Version, nil
		}
	}
	return "unknown", fmt.Errorf("could not find Kustomize version in build info: %+v", info)
}
```

If I (accidentally) try to `make WHAT=cmd/kubectl` with go 1.17, `debug.ReadBuildInfo()` returns "not ok". Is it acceptable at this point to have library code like this not work with versions < 1.18?

If I then fix my mistake and use 1.18, I end up at the second error condition, and it turns out `info.Deps` is nil: 

```go
&debug.BuildInfo{
  GoVersion:"go1.18", 
  Path:"k8s.io/kubernetes/cmd/kubectl", 
  Main:debug.Module{
    Path:"", Version:"", Sum:"", Replace:(*debug.Module)(nil)
  }, 
Deps:[]*debug.Module(nil), 
Settings:[]debug.BuildSetting{
  // tons of stuff, not about deps of course
}
```

If I try something very similar in the Kustomize repo, it works just fine (though is not available in tests, per [this issue](https://github.com/golang/go/issues/33976)). Consequently, I suspect the problem has something to do with this repo's build processes, which I admit I'm not familiar with in the slightest.

If I'm not doing something obviously wrong here and it doesn't "just work", does migrating to that strategy need to block this feature?

Reply:
> Is it acceptable at this point to have library code like this not work with versions < 1.18?

Once we're solidly on go1.18 (just switched this week and there are some follow up issues to work out), yes

> I suspect the problem has something to do with this repo's build processes

Yeah, I bet it's because kubernetes still builds with GOPATH instead of go modules. Work to switch over is in progress, but doesn't have a firm landing date.

> does migrating to that strategy need to block this feature?

I really don't think we should add build flags for deps this way. I would take much uglier solutions to avoid touching build scripts over this. Something like putting the kustomize version in a string constant in the kubectl version code and adding matchers to build/dependencies.yaml to make sure go.mod and that string constant stay in sync until we get to go1.18 module building

Reply:
adding more build-flag-based mechanisms is increasing technical debt in a direction we know we don't want to use and don't want downstreams to integrate with.

how about kubectl tries to use debug/buildinfo, which will work automatically for any downstream building with go modules. it can fallback to a package constant which we can populate from `cmd/kubectl/kubectl.go`, where we *do* know the version we're building with

Reply:
Wouldn't that constant still constitute tech debt that needs to be removed when go module support is enabled? I'm certainly not in favour of tech debt either, but I'm concerned a constant would just be a novel form of debt that we need to separately remember to clean up, still at the same time. In case it wasn't clear, I'm happy to tag myself on the TODO and clean it up personally (either way). 😄

Reply:
Ok, updated to use a constant. @soltysh preferred it to stay within staging, and is going to personally take care of oc.